File: sample/src/main/java/com/polidea/rxandroidble2/sample/example4_characteristic/advanced/PresenterEvent.java
package com.polidea.rxandroidble2.sample.example4_characteristic.advanced;


import androidx.annotation.NonNull;
import java.util.Arrays;

/**
 * A dummy interface to indicate what events may be emitted by the {@link Presenter}. This would be a whole lot more convenient
 * if used Kotlin's `sealed class` with additional `data class` subclasses.
 */
interface PresenterEvent {

}

class InfoEvent implements PresenterEvent {

    @NonNull final String infoText;

    InfoEvent(@NonNull String infoText) {
        this.infoText = infoText;
    }

    @Override
    public String toString() {
        return "InfoEvent{"
                + "infoText='" + infoText + '\''
                + '}';
    }
}

class ResultEvent implements PresenterEvent {

    @NonNull final Type type;

    @NonNull final byte[] result;

    ResultEvent(@NonNull byte[] result, @NonNull Type type) {
        this.result = result;
        this.type = type;
    }

    @Override
    public String toString() {
        return "ResultEvent{"
                + "type=" + type
                + ", result=" + Arrays.toString(result)
                + '}';
    }
}

class ErrorEvent implements PresenterEvent {

    @NonNull final Type type;

    @NonNull final Throwable error;

    ErrorEvent(@NonNull Throwable error, @NonNull Type type) {
        this.error = error;
        this.type = type;
    }

    @Override
    public String toString() {
        return "ErrorEvent{"
                + "type=" + type
                + ", error=" + error
                + '}';
    }
}

class CompatibilityModeEvent implements PresenterEvent {

    final boolean show;

    CompatibilityModeEvent(boolean show) {
        this.show = show;
    }

    @Override
    public String toString() {
        return "CompatibilityModeEvent{"
                + "show=" + show
                + '}';
    }
}

enum Type {
    READ, WRITE, NOTIFY, INDICATE
}

File: sample/src/main/java/com/polidea/rxandroidble2/sample/example4_characteristic/advanced/PresenterEvent.kt
package com.polidea.rxandroidble2.sample.example4_characteristic.advanced

/**
 * A dummy sealed class hierarchy to indicate what events may be emitted by the [Presenter].
 */
internal sealed class PresenterEvent

internal data class InfoEvent(val infoText: String) : PresenterEvent()

// Can't be a data class because of the ByteArray
internal class ResultEvent(val result: ByteArray, val type: Type) : PresenterEvent()

internal data class ErrorEvent(val error: Throwable, val type: Type) : PresenterEvent()

internal data class CompatibilityModeEvent(val show: Boolean) : PresenterEvent()

internal enum class Type {
    READ, WRITE, NOTIFY, INDICATE
}